stdlib/buffer.gr

```diff
@@ -0,0 +1,463 @@
+import Memory from "runtime/unsafe/memory"
+import WasmI32 from "runtime/unsafe/wasmi32"
+// import WasmI64 from "runtime/unsafe/wasmi64"
+// import WasmF32 from "runtime/unsafe/wasmf32"
+// import WasmF64 from "runtime/unsafe/wasmf64"
+import Conv from "runtime/unsafe/conv"
+// import { tagSimpleNumber, allocateBytes, allocateString } from "runtime/dataStructures"
+import Exception from "runtime/exception"
+import Int32 from "int32"
+import Bytes from "bytes"
+import String from "string"
+
+record Buffer {
+  mut size: Number,
+  initialSize: Number,
+  mut data: Bytes,
+}
+
+let _SIZE_OFFSET = 4n;
+
+let _VALUE_OFFSET = 8n;
+
+/** Converts Number to WasmI32 */
+@disableGC
+let numberToWasmI32 = n => {
+  let m = WasmI32.fromGrain(n)
+  if (WasmI32.eqz(WasmI32.and(m, 1n))) {
+    let i32 = Int32.fromNumber(n)
+    let res = Conv.fromInt32(i32)
+    Memory.free(WasmI32.fromGrain(i32))
+    res
+  } else {
+    // fast path
+    WasmI32.shrS(m, 1n)
+  }
+}
+
+/** Gets the size of a Bytes via its ptr */
+let getSize = (ptr) => WasmI32.load(ptr, _SIZE_OFFSET)
+
+let checkIndexIsInBounds = (i, len, buf) => {
+  if (i < 0) {
+    throw Exception.IndexOutOfBounds
+  }
+  if ((i + len) >= buf.size) {
+    throw Exception.IndexOutOfBounds
+  }
+}
+
+let grow = (buf) => {
+  buf.data = Bytes.resize(0, Bytes.length(buf.data), buf.data)
+}
+
+let autogrow = (len, buf) => {
+  if (buf.size + len >= Bytes.length(buf.data)) {
+    grow(buf)
+  }
+}
+
+let getCharAsWasmI32 = (char) => {
+  let c = WasmI32.fromGrain(char)
+  WasmI32.load8U(c, 4n)
+}
+
+let getCharByteLength = (byte) => {
+  let (+) = WasmI32.add
+  let (&) = WasmI32.and
+  let (==) = WasmI32.eq
+  if ((byte & 0x80n) == 0x00n) {
+    1n
+  } else if ((byte & 0xF0n) == 0xF0n) {
+    4n
+  } else if ((byte & 0xE0n) == 0xE0n) {
+    3n
+  } else {
+    2n
+  }
+}
+
+@disableGC
+let appendBytes = (srcOff, dstOff, len, src, dst) => {
+  let (+) = WasmI32.add
+  Memory.copy(dst + _VALUE_OFFSET + dstOff, src + _VALUE_OFFSET + srcOff, len)
+}
+
+// 8-bit
+// -----------------------------------------------------------------------------
+
+let _8BIT_LEN = 1
+
+export let getInt8S = (i, buf) => {
+  checkIndexIsInBounds(i, _8BIT_LEN, buf)
+  Bytes.getInt8S(i, buf.data)
+}
+
+export let getInt8U = (i, buf) => {
+  checkIndexIsInBounds(i, _8BIT_LEN, buf)
+  Bytes.getInt8U(i, buf.data)
+}
+
+export let setInt8 = (i, value, buf) => {
+  checkIndexIsInBounds(i, _8BIT_LEN, buf)
+  Bytes.setInt8(i, value, buf.data)
+}
+
+export let addInt8 = (value, buf) => {
+  checkIndexIsInBounds(buf.size, _8BIT_LEN, buf)
+  setInt8(buf.size, value, buf)
+}
+
+// 16-bit
+// -----------------------------------------------------------------------------
+
+let _16BIT_LEN = 2
+
+export let getInt16S = (i, buf) => {
```

Comment:
I'd love to see support for both little-endian and big-endian ints in the `Buffer` module (perhaps requires support from the `Bytes` module).

Reply:
That's fine, but it will likely be a breaking change. Additionally, we should document in this module the endianness of the operations in this PR, even if we don't support both.